src/h3lib/lib/h3Index.c

```diff
@@ -78,49 +78,108 @@ void H3_EXPORT(h3ToString)(H3Index h, char *str, size_t sz) {
+    // Implementation strategy:
+    //
+    // Walk from high to low bits, checking validity of
+    // groups of bits as we go. After each check, shift bits off
+    // to the left, so that the next relevant group is at the
+    // highest bit location in the H3 Index, which we can
+    // easily read off with the `GT` macro. This strategy helps
+    // us avoid re-computing shifts and masks for each group.
+    //
+    // |   Region   | # bits |
+    // |------------|--------|
+    // | High       |      1 |
+    // | Mode       |      4 |
+    // | Reserved   |      3 |
+    // | Resolution |      4 |
+    // | Base Cell  |      7 |
+    // | Digit 1    |      3 |
+    // | Digit 2    |      3 |
+    // | ...        |    ... |
+    // | Digit 15   |      3 |
+    //
+    // Additionally, we try to group operations and void loops when possible.
+
+    // The 1 high bit should be 0b0
+    // The 4 mode bits should be 0b0001 (H3_CELL_MODE)
+    // The 3 reserved bits should be 0b000
+    // In total, the top 8 bits should be 0b00001000
+    if (GT(h, 8) != 0b00001000) return 0;
+    h <<= 8;
+
+    // Number of bits in each of the resolution, base cell, and digit regions.
+    const int nBitsRes = 4;
+    const int nBitsBaseCell = 7;
+    const int nBitsDigit = H3_PER_DIGIT_OFFSET;
+
+    // No need to check resolution; any 4-bit number is a valid resolution.
+    const int res = GT(h, nBitsRes);
+    h <<= nBitsRes;
+
+    // Check that base cell number is valid.
+    const int bc = GT(h, nBitsBaseCell);
+    if (bc >= NUM_BASE_CELLS) return 0;
+    h <<= nBitsBaseCell;
+
+    // Now check that each resolution digit is valid.
+    // Let `r` denote the resolution we're currently checking.
+    int r = 1;
+
+    // Pentagon cells start with a sequence of 0's (CENTER_DIGIT's).
+    // The first nonzero digit can't be a 1 (i.e., "deleted subsequence",
+    // PENTAGON_SKIPPED_DIGIT, or K_AXES_DIGIT).
+    // Test for pentagon base cell first to avoid this loop if possible.
+    if (isBaseCellPentagonArr[bc]) {
+        for (; r <= res; r++) {
+            int d = GT(h, nBitsDigit);
+            if (d == 0) {
+                h <<= nBitsDigit;
+            } else if (d == 1) {
                 return 0;
+            } else {
+                break;
+                // But don't increment `r`, since we still need to
+                // check that it isn't a 7.
             }
         }
-
-        if (digit < CENTER_DIGIT || digit >= NUM_DIGITS) return 0;
     }
 
-    for (int r = res + 1; r <= MAX_H3_RES; r++) {
-        Direction digit = H3_GET_INDEX_DIGIT(h, r);
-        if (digit != INVALID_DIGIT) return 0;
+    // After (possibly) taking care of pentagon logic, check that
+    // the remaining digits up to `res` are not 7 (INVALID_DIGIT).
+    // Don't see a way to avoid this loop :(
+    for (; r <= res; r++) {
+        if (GT(h, nBitsDigit) == 7) return 0;
+        h <<= nBitsDigit;
```

Comment:
Assuming bitshifts are as expensive as they were when I was in college, it may make sense to just have an array of 16 masks and do `if (h & masks[r] == vals[r]) return 0;` as the only part of the inner loop.

Reply:
I ran some toy timing tests and didn't see much difference between masking vs shifting: https://gist.github.com/ajfriend/32571f0af1f3ea3133b6836fc861c730

Reply:
I wouldn't be surprised if both code generate the same (or almost) assembly behind.

Compiler are quite clever for this kind of simple expression nowadays, like you don't have to write `n >> 1` instead of `n / 2` and so on.
This is basic peephole optimisation.